doc/source/serve/production-guide/fault-tolerance.md

```diff
@@ -0,0 +1,273 @@
+You **must** deploy your Serve application with [KubeRay] to use this feature.
+
+See Serve's [Kubernetes production guide](serve-in-production-kubernetes) to learn how you can deploy your app with KubeRay.
+:::
+
+By default, Serve can recover from certain failures, such as unhealthy actors. When [Serve runs on Kubernetes](serve-in-production-kubernetes) with [KubeRay], it can also recover from some cluster-level failures, such as dead worker or head nodes.
+
+When a worker node fails, the actors running on it also fail. Serve detects that the actors have failed, and it attempts to respawn the actors on the remaining, healthy nodes. Meanwhile, KubeRay detects that the node itself has failed, so it brings up a new healthy node to replace it. Serve can then respawn any pending actors on that node as well. The deployment replicas running on healthy nodes can continue serving traffic throughout the recovery period.
+
+(serve-e2e-ft-guide-gcs)=
+### Head node recovery: Ray GCS fault tolerance
+
+:::{admonition} KubeRay Required
+:class: caution, dropdown
+You **must** deploy your Serve application with [KubeRay] to use this feature.
+
+See Serve's [Kubernetes production guide](serve-in-production-kubernetes) to learn how you can deploy your app with KubeRay.
+:::
+
+In this section, you'll learn how to add fault tolerance to Ray's Global Control Store (GCS), which allows your Serve application to serve traffic even when the head node crashes.
+
+By default the Ray head node is a single point of failure: if it crashes, the entire Ray cluster crashes and must be restarted. When running on Kubernetes, the `RayService` controller health-checks the Ray cluster and restarts it if this occurs, but this introduces some downtime.
+
+In Ray 2.0, KubeRay has added experimental support for [Global Control Store (GCS) fault tolerance](https://ray-project.github.io/kuberay/guidance/gcs-ft/#ray-gcs-fault-tolerancegcs-ft-experimental), preventing the Ray cluster from crashing if the head node goes down.
+While the head node is recovering, Serve applications can still handle traffic but cannot be updated or recover from other failures (e.g. actors or worker nodes crashing).
+Once the GCS is recovered, the cluster will return to normal behavior.
+
+You can enable GCS fault tolerance on KubeRay by adding an external Redis server and modifying your `RayService` Kubernetes object.
+
+Below, we explain how to do each of these.
+
+#### Step 1: Add external Redis server
+
+GCS fault tolerance requires an external Redis database. You can choose to host your own Redis database, or you can use one through a third-party vendor. We recommend using a highly-available Redis database for resiliency.
+
+**For development puposes**, you can also host a Redis database on the same Kubernetes cluster as your Ray cluster. For example, you can add a 1-node Redis cluster by prepending these three Redis objects to your Kubernetes YAML:
+
+(one-node-redis-example)=
```

Comment:
refer the yaml file section?

Reply:
Should I link to the KubeRay docs? AFAIK, none of the Serve docs shows a full Kubernetes YAML.

Reply:
Oh I see– I didn't link it because I just wanted to show the redis objects in this section. The full yaml also includes the Serve config. I can't guard it with a comment because the comment will show up when I refer to the full yaml later. Any ideas on how to get around this?